### ISSUE TITLE
Unable to run kong in docker swarm mode

### ISSUE THREAD

**Issue by michaelkrog** [Kong/kong#1981]
### Summary

I have been trying to get Kong running in Docker Swarm mode. But it does not seem to start properly. It does not respond.

### Steps To Reproduce

1. Create a docker swarm (mode).
2. Create an overlay network `docker network create --driver overlay my-net`
3. Create a cassandra service `docker service create --network my-net --name cassandra01 -e CASSANDRA_LISTEN_ADDRESS=cassandra01 cassandra:2.2`
4. Create a kong service `docker service create --network my-net --name kong -e KONG_CASSANDRA_CONTACT_POINTS=cassandra01 -e KONG_LOG_LEVEL=debug -e KONG_DATABASE=cassandra -p 8001:8001 kong`
5. Test that kong does not respond `curl http://127.0.0.1:8001`

### Additional Details & Logs

Only log at startup is this:
```
2017/01/13 09:30:12 [warn] 7#0: *2 [lua] log.lua:22: warn(): No cluster infos in shared dict cassandra, context: ngx.timer
```

The internal health looks like this:
```
$ docker exec -ti kong.1.abpqhdxzwllu318kga2cwpjqj kong health
serf........not running
dnsmasq.....not running
nginx.......not running

Error: Kong is not running at /usr/local/kong
```

**Comment by studyingMole** [comment]
I've got the same problem. Tried on Kubernetes and got similar error as: `[lua] log.lua:22: warn(): No cluster infos in shared dict cassandra, context: ngx.timer`

Then tried to launch local Docker containers and got the same error. Steps were:

1. `docker run -d -p 9042:9042 --name api-db cassandra:2.2.8`
2. `docker run -d --name api-gw --link api-db:api-db -e "KONG_DATABASE=cassandra" -e "KONG_CASSANDRA_CONTACT_POINTS=api-db" -p 8000:8000 -p 8443:8443 -p 8001:8001 -p 7946:7946 -p 7946:794
6/udp kong`
3. `docker logs api-gw`

Logs returns the error: `2017/01/16 11:21:21 [warn] 7#0: *2 [lua] log.lua:22: warn(): No cluster infos in shared dict cassandra, context: ngx.timer`

**Comment by studyingMole** [comment]
Tried again with the same steps but used Docker tag `kong:0.9.6` which worked! 

So, on step 2: `docker run -d --name api-gw --link api-db:api-db -e "KONG_DATABASE=cassandra" -e "KONG_CASSANDRA_CONTACT_POINTS=api-db" -p 8000:8000 -p 8443:8443 -p 8001:8001 -p 7946:7946 -p 7946:7946/udp kong:0.9.6` runs without any errors.

I can successfully load an API using `kong:0.9.6`, so might be something wrong with the `0.9.7` docker image?

**Comment by thefosk** [comment]
As far as I know the following log can be ignored:

```
2017/01/13 09:30:12 [warn] 7#0: *2 [lua] log.lua:22: warn(): No cluster infos in shared dict cassandra, context: ngx.timer
```

But @thibaultcha can explain why it is being displayed. I will let @shashiranjan84 debug the issue.

**Comment by shashiranjan84** [comment]
@michaelkrog @studyingMole I just tested Kong:0.9.8 and it worked for me. Can you try with latest version of Kong?

**Comment by shashiranjan84** [comment]
@michaelkrog is Cassandra:2.2 cluster aware? Not sure it will work properly on a cluster.

**Comment by shashiranjan84** [comment]
@studyingMole Even Kong:0.9.7 worked fine for me.

**Comment by ArunParthiban10** [comment]
@shashiranjan84, I'm using a Cassandra cluster and Kong:0.9.8.  And I'm getting 
2017/01/24 17:02:37 [warn] 7#0: *2 [lua] log.lua:22: warn(): No cluster infos in shared dict cassandra, context: ngx.timer

Is Cassandra clustering unsupported atm ?

**Comment by thibaultcha** [comment]
> 2017/01/24 17:02:37 [warn] 7#0: *2 [lua] log.lua:22: warn(): No cluster infos in shared dict cassandra, context: ngx.timer

This logs means that the driver will actually proceed to refreshing the Cassandra cluster's state. It won't be updated since it doesn't exit anymore in the most recent version of the driver. Enable `info` logs if you wish to see more details about the fetched cluster's state.

**Comment by CalSmith1** [comment]
I get this error on Kong 0.9.9.
The next line in my log says:
"could not start Kong, stopping services"
is the "no cluster infos..." warning keeping Kong from starting?
TIA

**Comment by shashiranjan84** [comment]
@CalSmith1 No you can safely ignore that error, log into container for error logs.

**Comment by CalSmith1** [comment]
This was actually a Cassandra incompatible release problem.
I had inadvertently installed Cassandra 3.x instead of 2.x.
The resulting errors lead me to this post - Thanks for your help

**Comment by michaelkrog** [comment]
@shashiranjan84 I do get a bit more info now:
```
2017/02/09 21:11:30 [warn] 7#0: *2 [lua] log.lua:22: warn(): No cluster infos in shared dict cassandra, context: ngx.timer
2017/02/09 21:11:30 [warn] 7#0: *2 [lua] log.lua:22: warn(): No cluster infos in shared dict cassandra, context: ngx.timer
Error: /usr/local/share/lua/5.1/kong/cmd/start.lua:18: [cassandra error] All hosts tried for query failed. cassandra01: connection refused for socket with peer cassandra01.
```

So it seems it have difficulty connecting to cassandra. The weird thing is that when I look into cassandra logs I get stuff like this:
```
INFO  21:12:10 Initializing kong.oauth2_tokens.oauth2_tokens_authenticated_userid_idx
INFO  21:12:10 Writing Memtable-IndexInfo@1447077371(0.066KiB serialized bytes, 1 ops, 0%/0% of on/off-heap limit)
INFO  21:12:11 Update table 'kong/oauth2_tokens' From org.apache.cassandra.config.CFMetaData@5f91df01[cfId=6a48a2d0-ef0c-11e6-b914-bb96bfb5171a,ksName=kong,cfName=oauth2_tokens,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.UTF8Type),comment=,readRepairChance=0.0,dcLocalReadRepairChance=0.1,gcGraceSeconds=864000,defaultValidator=org.apache.cassandra.db.marshal.BytesType,keyValidator=org.apache.cassandra.db.marshal.UUIDType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=id, type=org.apache.cassandra.db.marshal.UUIDType, kind=PARTITION_KEY, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=created_at, type=org.apache.cassandra.db.marshal.TimestampType, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=token_type, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=scope, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=refresh_token, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_tokens_refresh_token_idx, indexType=COMPOSITES}, ColumnDefinition{name=credential_id, type=org.apache.cassandra.db.marshal.UUIDType, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=authenticated_userid, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_tokens_authenticated_userid_idx, indexType=COMPOSITES}, ColumnDefinition{name=access_token, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_tokens_access_token_idx, indexType=COMPOSITES}, ColumnDefinition{name=expires_in, type=org.apache.cassandra.db.marshal.Int32Type, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}],compactionStrategyClass=class org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy,compactionStrategyOptions={},compressionParameters={sstable_compression=org.apache.cassandra.io.compress.LZ4Compressor},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL", "rows_per_partition":"NONE"},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,droppedColumns={},triggers=[],isDense=false] To org.apache.cassandra.config.CFMetaData@6676efd[cfId=6a48a2d0-ef0c-11e6-b914-bb96bfb5171a,ksName=kong,cfName=oauth2_tokens,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.UTF8Type),comment=,readRepairChance=0.0,dcLocalReadRepairChance=0.1,gcGraceSeconds=864000,defaultValidator=org.apache.cassandra.db.marshal.BytesType,keyValidator=org.apache.cassandra.db.marshal.UUIDType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=created_at, type=org.apache.cassandra.db.marshal.TimestampType, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=id, type=org.apache.cassandra.db.marshal.UUIDType, kind=PARTITION_KEY, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=token_type, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=scope, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=authenticated_userid, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_tokens_authenticated_userid_idx, indexType=COMPOSITES}, ColumnDefinition{name=credential_id, type=org.apache.cassandra.db.marshal.UUIDType, kind=REGULAR, componentIndex=0, indexName=oauth2_credential_id_idx, indexType=COMPOSITES}, ColumnDefinition{name=refresh_token, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_tokens_refresh_token_idx, indexType=COMPOSITES}, ColumnDefinition{name=access_token, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_tokens_access_token_idx, indexType=COMPOSITES}, ColumnDefinition{name=expires_in, type=org.apache.cassandra.db.marshal.Int32Type, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}],compactionStrategyClass=class org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy,compactionStrategyOptions={},compressionParameters={sstable_compression=org.apache.cassandra.io.compress.LZ4Compressor},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL", "rows_per_partition":"NONE"},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,droppedColumns={},triggers=[],isDense=false]
INFO  21:12:11 Writing Memtable-schema_keyspaces@670094393(0.143KiB serialized bytes, 3 ops, 0%/0% of on/off-heap limit)
INFO  21:12:11 Writing Memtable-schema_columnfamilies@291120478(1.272KiB serialized bytes, 23 ops, 0%/0% of on/off-heap limit)
INFO  21:12:11 Writing Memtable-schema_columns@1077977443(0.385KiB serialized bytes, 7 ops, 0%/0% of on/off-heap limit)
INFO  21:12:11 Initializing kong.oauth2_tokens.oauth2_credential_id_idx
INFO  21:12:11 Writing Memtable-IndexInfo@1119003486(0.053KiB serialized bytes, 1 ops, 0%/0% of on/off-heap limit)
INFO  21:12:12 Update table 'kong/oauth2_authorization_codes' From org.apache.cassandra.config.CFMetaData@5e66a53[cfId=68deb290-ef0c-11e6-b914-bb96bfb5171a,ksName=kong,cfName=oauth2_authorization_codes,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.UTF8Type),comment=,readRepairChance=0.0,dcLocalReadRepairChance=0.1,gcGraceSeconds=864000,defaultValidator=org.apache.cassandra.db.marshal.BytesType,keyValidator=org.apache.cassandra.db.marshal.UUIDType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=id, type=org.apache.cassandra.db.marshal.UUIDType, kind=PARTITION_KEY, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=created_at, type=org.apache.cassandra.db.marshal.TimestampType, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=scope, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=authenticated_userid, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_authorization_codes_authenticated_userid_idx, indexType=COMPOSITES}, ColumnDefinition{name=code, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_authorization_codes_code_idx, indexType=COMPOSITES}],compactionStrategyClass=class org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy,compactionStrategyOptions={},compressionParameters={sstable_compression=org.apache.cassandra.io.compress.LZ4Compressor},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL", "rows_per_partition":"NONE"},defaultTimeToLive=300,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,droppedColumns={},triggers=[],isDense=false] To org.apache.cassandra.config.CFMetaData@7e228be9[cfId=68deb290-ef0c-11e6-b914-bb96bfb5171a,ksName=kong,cfName=oauth2_authorization_codes,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.UTF8Type),comment=,readRepairChance=0.0,dcLocalReadRepairChance=0.1,gcGraceSeconds=864000,defaultValidator=org.apache.cassandra.db.marshal.BytesType,keyValidator=org.apache.cassandra.db.marshal.UUIDType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=created_at, type=org.apache.cassandra.db.marshal.TimestampType, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=id, type=org.apache.cassandra.db.marshal.UUIDType, kind=PARTITION_KEY, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=scope, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=null, indexType=null}, ColumnDefinition{name=authenticated_userid, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_authorization_codes_authenticated_userid_idx, indexType=COMPOSITES}, ColumnDefinition{name=code, type=org.apache.cassandra.db.marshal.UTF8Type, kind=REGULAR, componentIndex=0, indexName=oauth2_authorization_codes_code_idx, indexType=COMPOSITES}],compactionStrategyClass=class org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy,compactionStrategyOptions={},compressionParameters={sstable_compression=org.apache.cassandra.io.compress.LZ4Compressor},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL", "rows_per_partition":"NONE"},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,droppedColumns={},triggers=[],isDense=false]
INFO  21:12:12 Writing Memtable-schema_keyspaces@1680280872(0.143KiB serialized bytes, 3 ops, 0%/0% of on/off-heap limit)
INFO  21:12:12 Writing Memtable-schema_columnfamilies@309493413(1.564KiB serialized bytes, 23 ops, 0%/0% of on/off-heap limit)
```

So apparently I has something going on with cassandra, but still spits our connection errors. :-/

**Comment by shashiranjan84** [comment]
@michaelkrog  Kong is not able to reach Cassandra instance `cassandra01`. Are you able to telnet/nc to cassandra instance from Kong container?

**Comment by christtrc** [comment]
I use Cassandra 3.10, and I got this error:
```
2017/02/10 11:00:12 [warn] 7#0: *2 [lua] log.lua:22: warn(): No cluster infos in shared dict cassandra, context: ngx.timer
2017/02/10 11:00:12 [error] 7#0: *2 [lua] log.lua:22: err(): Could not set keyspace. [Invalid] Keyspace 'kong' does not exist, context: ngx.timer
Error: /usr/local/share/lua/5.1/kong/cmd/start.lua:18: [cassandra error] [Invalid] unconfigured table schema_keyspaces
```

**Comment by Tieske** [comment]
@christtrc Please use Cassandra 2.2, version 3 is only supported from Kong 0.10

**Comment by christtrc** [comment]
OK, Thanks! I saw there was a 0.10rc3 tag, whnen the final version will come?

**Comment by Tieske** [comment]
rc4 shortly, if that remains without bugs, final 2 weeks later

**Comment by michaelkrog** [comment]
I just tried again with version 0.10 and still have similar issues.

What bothers me is that Kong can connect and create tables in Cassandra just fine, but then afterwards it seems unable to connect.

(on mobile right now, will supply more info soon)

**Comment by michaelkrog** [comment]
However...

If I create an [attachable](https://docs.docker.com/engine/reference/commandline/network_create/) overlay network and run ordinary containers connected to the overlay network, then is works fine.

Something goes wrong as soon as I use docker services with kong. :(

**Comment by Tieske** [comment]
@michaelkrog only read the last posts, two things that come to mind;

 - use fully qualified names for cassandra db
 - make sure only 1 container starts and runs the migrations. When that is done, start the others.

**Comment by JnMik** [comment]
I was gonna type that I have the same EXACT issue as @michaelkrog .. and then while typing I found the error.

My cassandra_contact_points config in kong.conf wasn't separated by coma but with semi-colon. 
Fixed it and now my first kong node seems to boot and pass beyond that error, we will see If I can get to a fully working setup.

**Comment by mrgavinconway001** [comment]
Would be interested to know if you got this working under swarm. Could you share your docker service definition?

**Comment by michaelkrog** [comment]
I tried with a newer version later on which worked without any issues.

Next issue was to get  it working in as a cluster. That required that I changed endpoint-mode from vip to dnsrr. That way the containers can reach each other by ip.

**Comment by thefosk** [comment]
@michaelkrog what version of Kong did you use that fixed this issue?

**Comment by michaelkrog** [comment]
@thefosk Currently we are running 0.10.2 using Cassandra backend. It works for us.

**Comment by endeepak** [comment]
We're facing issues running kong cluster in docker swarm. The cluster API returns different number of kong nodes in different node. I've raised this issue in https://github.com/Mashape/docker-kong/issues/93

**Comment by thefosk** [comment]
Please note that in 0.11 Serf has been removed so that won't be a dependency anymore, and Kong nodes don't have to cluster together anymore.

**Comment by ctessier** [comment]
Any update on this issue?

I am myself using Kong 0.12.2 and Cassandra 3.11 and run into this error while trying to run the migrations:

>  Error: [cassandra error] Error during migration 2016-07-15-oauth2_code_credential_id: [Truncate error] Error during truncate: Truncate timed out - received only 0 responses 

Thank you for your help :)

**Comment by supritshah1289** [comment]
@michaelkrog @ctessier Were you able to solve this issue? 
1. I have created overlay network with --attachable
2. Docker stack deploy a single instance of Cassandra on a docker swarm cluster.
3. Run migration on a cluster through docker stack deploy. (I have also tried running the migration on running cassandra service hence overlay network is created with --attachable)

After performing all the steps listed above, I am getting the same error. Below is the error I ran into while running the migration. 
```
migrating oauth2 for keyspace kong
Error: [cassandra error] Error during migration 2016-07-15-oauth2_code_credential_id: [Truncate error] Error during truncate: Truncate timed out - received only 0 responses
```

docker version: Docker version 18.03.1-ce, build 9ee9f40
Server: Red Hat Linux 4.8
Kong: 0.14
Cassandra: 3.0

**Comment by p0pr0ck5** [comment]
@shashiranjan84 if you've time, are you able to eyeball this? You're assigned to it, and it seems pretty stale, but I'm not sure if there's anything actionable yet here.

**Comment by shashiranjan84** [comment]
Problem seems to be with Cassandra than Kong, more of networking issue on overlay network. @supritshah1289 are you still facing the issue?

**Comment by ctessier** [comment]
> @michaelkrog @ctessier Were you able to solve this issue?
> 
> 1. I have created overlay network with --attachable
> 2. Docker stack deploy a single instance of Cassandra on a docker swarm cluster.
> 3. Run migration on a cluster through docker stack deploy. (I have also tried running the migration on running cassandra service hence overlay network is created with --attachable)
> 
> After performing all the steps listed above, I am getting the same error. Below is the error I ran into while running the migration.
> 
> ```
> migrating oauth2 for keyspace kong
> Error: [cassandra error] Error during migration 2016-07-15-oauth2_code_credential_id: [Truncate error] Error during truncate: Truncate timed out - received only 0 responses
> ```
> docker version: Docker version 18.03.1-ce, build 9ee9f40
> Server: Red Hat Linux 4.8
> Kong: 0.14
> Cassandra: 3.0

Hi @supritshah1289, I didn't reproduce the issue since last july. Since we are not using the cluster mode, we switched to Postgres.

**Comment by supritshah1289** [comment]
@shashiranjan84  Yes, I am still facing the same issues.

**Comment by thibaultcha** [comment]
@supritshah1289 Have you tried with Kong 1.0? The latest versions of Kong include a fix in the Cassandra driver for some of the error messages you encountered here.

**Comment by supritshah1289** [comment]
@thibaultcha I have not. I will give try it on Monday. Thank you

**Comment by subnetmarco** [comment]
Are there any updates on this issue?

**Comment by JnMik** [comment]
I managed to have one up and running last year, inside a swarm cluster. But the sequence to boot up the stack was pretty slow because cassandra nodes had to wait for the previous one to be ready before booting, I believe I also had to spawn the first kong container alone so it can run migrations, and the others after. 

It was a bit of a pain, but I got it working. Still in the end I found it simpler to run it on VMs fully provisioned with packer and drop the containers. It was more stable, containers was not moving around nodes, issues with cassandra cluster health was gone, boot speed was increased and more .. 

Still curious to know if anyone run this with swarm in production without a sweat ?
Following !
